extensions/ql-vscode/src/remote-queries/variant-analysis-manager.ts

```diff
@@ -56,22 +59,42 @@ export class VariantAnalysisManager extends DisposableObject implements VariantA
     this.variantAnalysisResultsManager.onResultLoaded(this.onRepoResultLoaded.bind(this));
   }
 
-  public async rehydrateVariantAnalysis(variantAnalysis: VariantAnalysis, status: QueryStatus) {
+  public async rehydrateVariantAnalysis(variantAnalysis: VariantAnalysis) {
     if (!(await this.variantAnalysisRecordExists(variantAnalysis.id))) {
       // In this case, the variant analysis was deleted from disk, most likely because
       // it was purged by another workspace.
       this._onVariantAnalysisRemoved.fire(variantAnalysis);
     } else {
       this.variantAnalyses.set(variantAnalysis.id, variantAnalysis);
       await this.getView(variantAnalysis.id)?.updateView(variantAnalysis);
-      if (status === QueryStatus.InProgress) {
-        // In this case, last time we checked, the query was still in progress.
-        // We need to setup the monitor to check for completion.
+
+      if (!await this.isVariantAnalysisComplete(variantAnalysis)) {
         await commands.executeCommand('codeQL.monitorVariantAnalysis', variantAnalysis);
       }
     }
   }
 
+  public async isVariantAnalysisComplete(variantAnalysis: VariantAnalysis): Promise<boolean> {
+    // It's only acceptable to have no scanned repos if the variant analysis is not in a final state.
+    // Otherwise it means the analysis hit some kind of internal error or there were no repos to scan.
+    if (variantAnalysis.scannedRepos === undefined || variantAnalysis.scannedRepos.length === 0) {
+      return variantAnalysis.status !== VariantAnalysisStatus.InProgress;
+    }
+
+    return (await Promise.all(variantAnalysis.scannedRepos.map(repo => this.isVariantAnalysisRepoComplete(variantAnalysis.id, repo)))).every(x => x);
+  }
+
+  private async isVariantAnalysisRepoComplete(variantAnalysisId: number, repo: VariantAnalysisScannedRepository): Promise<boolean> {
+    if (!hasRepoScanCompleted(repo)) {
+      return false;
+    } else if (!repoScanHasResults(repo)) {
+      return true;
+    } else {
+      const storageLocation = this.getVariantAnalysisStorageLocation(variantAnalysisId);
+      return await this.variantAnalysisResultsManager.isVariantAnalysisRepoDownloaded(storageLocation, repo.repository.fullName);
+    }
+  }
```

Comment:
Since this doesn't depend on any instance variables, I think you can extract this to a function which is outside of the class. We could have a simple function like `isVariantAnalysisComplete(variantAnalysisResultsManager: VariantAnalysisResultsManager, variantAnalysisStorageLocation: string, variantAnalysis: VariantAnalysis)` which can be called from anywhere. This would also solve the problem of this being unnecessarily `public` since we can just `export` it and not think about its visibility.

Reply:
Agree with @koesie10 - this function can be extracted and unit tested/mocked independently. 

If we want to avoid passing in the `VariantAnalysisResultsManager`, we could instead pass in a function that helps decide whether a repo has downloads.

Reply:
Thanks for the advice. That makes sense to me, to extract the method and pass in a function that checks whether the repo has downloads or not. This way we can keep the logic pure and contained to `variant-analysis.ts` (unless you have suggestions for somewhere better to put it). And we can test the logic of this method and the logic of rehydrating separately.